packages/state-transition/src/cache/pubkeyCache.ts

```diff
@@ -1,11 +1,13 @@
 import {CoordType, PublicKey} from "@chainsafe/bls/types";
 import bls from "@chainsafe/bls";
+import * as immutable from "immutable";
```

Comment:
> using the vanilla immutable implementation as it fits our needs at least API-wise. @wemeetagain @tuyennhv any consideration you want to mention on why we forked `persistent-ts` https://github.com/ChainSafe/ssz/tree/master/packages/persistent-ts ?

Reply:
as I remember, the main reason is it's typescript first, with more straightforward implementation and better performance (or at least comparable to immutablejs)

> This library isn't as developed as others, such as [immutable js](https://github.com/immutable-js/immutable-js). However, it can provide alternative data structures, and hopefully more readable implementations. Unlike immutable js specifically, this implementation is also Typescript first, whereas that library adds type annotations after the fact. This makes for an api centered around generics, and not necessarily acccomadating JS' indiosyncracies.

Reply:
also one of the capability of `MutableVector` in `persistent-ts` is it supports batching multiple `push()` `pop()` operations through `TransientVector` which mutates the tail of the underlying data structure, this helps it more efficient

for example, when we call `processDeposit` `n` times in the same block, we want to create one immutable object instead of `n` immutable objects and only use the last one. Not sure if `immutable.Map` supports this

Reply:
for the context, persistent-ts was a port of Clojure implementation to Typescript, and I think ImmutableJS also referred to that. I see their `PersistentHashMap` also support transient mode ("batch mutation") https://github.com/clojure/clojure/blob/master/src/jvm/clojure/lang/PersistentHashMap.java

ImmutableJS also support batching mutation https://immutable-js.com/#batching-mutations

Reply:
> for the context, persistent-ts was a port of Clojure implementation to Typescript, and I think ImmutableJS also referred to that. I see their `PersistentHashMap` also support transient mode ("batch mutation") https://github.com/clojure/clojure/blob/master/src/jvm/clojure/lang/PersistentHashMap.java
> 
> ImmutableJS also support batching mutation https://immutable-js.com/#batching-mutations

Do you think it is necessary to do batch mutation for unfinalized pubkey cache? 
Theoretically, each block contains up to 16 deposits, so it will call immutable.Map.set() at most 16 times. I am not sure how much difference of performance is there between doing set() 16 times vs batch mutation.

I went for former because it is easier to read, and more aligned with the spec ie. call `process_deposit()` 16 times

Reply:
Post eip6110 the number of deposits per block is no longer bounded. The economical max is ~700

Reply:
you should proof good performance of this cache, by adding performance tests and running some memory tests locally. Examples of memory tests are here, a bit imprecise but better than nothing https://github.com/ChainSafe/lodestar/tree/unstable/packages/beacon-node/test/memory